services/web-server/src/PulseEngine/PulseIterator.js

```diff
@@ -38,40 +44,65 @@ export default class PulseIterator {
     return this;
   }
 
+  // cause this to throw the given error for all pending and subsequent next() calls
+  pushError(error) {
+    this.stop({error, done: true});
+  }
+
+  // insert a value into the queue; the resulting promise resolves when the value
+  // is consumed, and rejects if the iterator finishes without consuming the value
   pushValue(value) {
-    if (this.pullQueue.size !== 0) {
-      this.pullQueue.first()({ value, done: false });
-      this.pullQueue = this.pullQueue.shift();
-    } else {
-      this.pushQueue = this.pushQueue.push(value);
+    if (!this.listening) {
+      return Promise.reject(new Error('iterator cancelled'));
     }
+    return new Promise((resolve, reject) => {
+      this.pushQueue = this.pushQueue.push({resolve, reject, value, done: false});
+      this.match();
+    });
   }
 
+  // pull a value from the iterator; the resulting promise resolves when a value
+  // is available or the iterator is finished, and rejects if the iterator encounters
+  // an error
   pullValue() {
-    return new Promise(resolve => {
-      if (this.pushQueue.size !== 0) {
-        resolve({ value: this.pushQueue.first(), done: false });
-        this.pushQueue = this.pushQueue.shift();
-      } else {
-        this.pullQueue = this.pullQueue.push(resolve);
-      }
+    return new Promise((resolve, reject) => {
+      this.pullQueue = this.pullQueue.push({resolve, reject});
+      this.match();
     });
   }
 
-  emptyQueue() {
-    if (!this.listening) {
-      return;
+  stop(push) {
+    if (this.listening) {
+      this.listening = false;
+      this.pulseEngine.unsubscribe(this.subscriptionId);
+
+      // reject any pending pushes and then replace with a single, persistent push
+      const err = new Error('iterator cancelled');
+      this.pushQueue.forEach(({reject}) => reject(err));
+      this.pushQueue = this.pushQueue.clear().push(push);
     }
+    this.match();
+  }
 
-    this.listening = false;
-    this.pulseEngine.unsubscribe(this.subscriptionId);
-    this.pullQueue.forEach(resolve =>
-      resolve({
-        value: undefined,
-        done: true,
-      })
-    );
-    this.pullQueue = this.pullQueue.clear();
-    this.pushQueue = this.pushQueue.clear();
+  match() {
+    while (this.pushQueue.size !== 0 && this.pullQueue.size !== 0) {
+      const push = this.pushQueue.first();
+      const pull = this.pullQueue.first();
+
+      this.pullQueue = this.pullQueue.shift();
+      if (!push.done) {
+        this.pushQueue = this.pushQueue.shift();
+      }
+
+      if (push.error) {
+        pull.reject(push.error);
+      } else {
+        pull.resolve({value: push.value, done: push.done});
```

Comment:
Oh, ok, so it seems we now always return `Promise` of `{value: blahblah, done: blahblah}`